proxy/coco/src/announce.rs

```diff
@@ -0,0 +1,303 @@
+//! Compute, track and announce noteworthy changes to the network.
+
+use std::collections::HashSet;
+use std::ops::Deref as _;
+
+use kv::Codec as _;
+
+use librad::keys;
+use librad::net;
+use librad::uri;
+
+use crate::error::Error;
+use crate::oid::Oid;
+use crate::peer;
+
+/// Name for the bucket used in [`kv::Store`].
+const BUCKET_NAME: &str = "announcements";
+/// Key for the single value used as cache.
+const KEY_NAME: &str = "latest";
+
+/// An update and all the required information that can be announced on the network.
+pub type Announcement = (uri::RadUrn, String, Oid);
+
+/// Announces the list of given `updates` with the [`librad::net::protocol`].
+///
+/// # Errors
+///
+/// * if the announcemnet of one of the project heads failed
+pub async fn announce(
+    protocol: net::protocol::Protocol<net::peer::PeerStorage<keys::SecretKey>, net::peer::Gossip>,
+    updates: Vec<Announcement>,
+) -> Result<(), Error> {
+    for (project_urn, head, hash) in &updates {
+        let urn = uri::RadUrn::new(
+            project_urn.id.clone(),
```

Comment:
Since we only use the `id` here, should `Announcement`  be `(Hash, String, Oid)`?

Reply:
Actually you should probably leave as-is, but use `urn.protocol` below.

Reply:
I think adding a `set_path`/`with_path` method to `RadUrn` would generally be useful.

Reply:
@FintanH Do we want to capture a follow-up for this?

Reply:
Can do :+1:

Reply:
Oh wait. The fields are public. I think you can do `RadUrn { path..old_urn }`